base/timing.jl

```diff
@@ -185,21 +189,52 @@ julia> @time begin
 """
 macro time(ex)
     quote
+        # First, measure how long it takes to compile the @time block without a body
+        local empty_body_compiletime = cumulative_compile_time_ns()
+        $(_time_expr(nothing))
+        empty_body_compiletime = cumulative_compile_time_ns() - empty_body_compiletime
+
+        # Then, run the real code, and measure the compilation time there, and subtract out
+        # the time for just the overhead.
+        local outer_compiletime = cumulative_compile_time_ns()
+        elapsedtime, inner_compiletime, gcdiff, val = $(_time_expr(ex))
+        outer_compiletime = cumulative_compile_time_ns() - outer_compiletime
+
+        # Subtract out the inner (dynamic dispatch) compilation time to get toplevel + overhead.
+        outer_compiletime = outer_compiletime - inner_compiletime
+        # Threshold to display compilation time (because computers have hella variance)
+        overhead_ratio = outer_compiletime / empty_body_compiletime
+        if overhead_ratio > 2.0  # three standard deviations of noise in our experiments
+            # Subtract out the overhead, so we reflect only the true compiletime for user code.
+            toplevel_compiletime = outer_compiletime - empty_body_compiletime
+        else
+            toplevel_compiletime = 0
+        end
+
+        time_print(elapsedtime, gcdiff.allocd, gcdiff.total_time,
+                   gc_alloc_count(gcdiff), inner_compiletime, toplevel_compiletime)
+
+        println()
+        val
+    end
+end
+
+function _time_expr(ex)
+    Expr(:toplevel, quote
```

Comment:
Does this change using `@time` in local scope?

Reply:
Oh, oops! Thanks @yuyichao! We worked really hard to avoid changing how `@time` behaves in a local scope (if we didn't care about that, this could have been much simpler).

We added this line at some point while debugging, and i meant to remove it. Hopefully the fact that it's working isn't pinned on this...

Reply:
Ah shit, indeed removing this does break the implementation 🙄 

Thanks for catching this! I'll see if it can be recovered, but your other comment might be right.